### PR TITLE
Python client

### PR DESCRIPTION
**Pull request by jmfrye**
First pass at Python 2.7 implementation of TreodeDB Client Library

### CODE REVIEW

**Comment by BrainFlame** [comment]
There should be no need for multiple .gitignore files.  .gitignore inside sub folders are to add types that you don't want to ignore in the global .gitignore. The global .gitignore seems to already have .pyc .

Python2.7 should come with functools, although in the README it is suggested to install it.
On the other hand, I received an ImportError: No module named transaction. It is not clear what to install to get 'transaction' module. I tried pip install transaction and got me a zope.transaction module that does not work as expected.

**Comment by jmfrye** [comment]
BrainFlame, I removed the extra .gitignore files.  Thanks.

I removed functools from the README.  Thanks again.

The transaction module is part of the client library.  I think the issue is that you cannot run the tests directly from the tests directory because they cannot find the modules in the parent folder.  That's a Python issue I don't know how to get around other than just run

python -m unittest discover 

instead.  Any better ideas?

**Comment by BrainFlame** [comment]
Found the issue. Please rename the file Transaction.py to transaction.py (all lowercase).

**Comment by jmfrye** [comment]
Done.  I had just renamed it locally before.

**Comment by Topher-the-Geek** [comment]
Please rebase and squash your changes from the latest master branch.

Add a copyright header to all the .py files. You can find the text in the .scala files; just convert it to Python's comment format.

File: client/python/tx_clock.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    # In Python, longs are actually unbounded, but we'll give TxClock a max
+    # TODO: How big do we want the max to be?
+    MaxValue = 2**64 - 1
+
+    # REQUIRES: Input time in seconds
+    def __init__(self, time=None, input_in_usecs=False):
+        self._input_in_usecs = input_in_usecs
+
+        # Default to current time
+        if (time == None):
```
Comment (Topher-the-Geek) [inline client/python/tx_clock.py]: Convert the system time to a long in the constructor, and drop the epsilon compare below.

File: client/python/tx_clock.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            # time in seconds
+            time = time.time()
+
+        # Convert time to microseconds if needed
+        self.u_factor = 6
+        if (self._input_in_usecs):
```
Comment (Topher-the-Geek) [inline client/python/tx_clock.py]: I don't understand what this is about.

File: client/python/transaction.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
+import time
+
+from cache_map import *
+from client_cache import *
+from tx_clock import *
+
+class Transaction(object):
+    def __init__(self, cache, read_timestamp=None, max_age=None, no_cache=False):
+        if (type(cache) != ClientCache):
```
Comment (Topher-the-Geek) [inline client/python/transaction.py]: I don't think you want to enforce this. In fact, I think you specifically want to pass in a mock of ClientCache for testing transaction.

File: client/python/tests/test_transaction.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            self.test_commit
+        ]
+
+    def test_read(self):
+        print "test_read",
+        server = "www.google.com"
+        cache = ClientCache(server)
+
+        # Mock the request method
+        headers = {
```
Comment (Topher-the-Geek) [inline client/python/tests/test_transaction.py]: Mocking HTTP is necessary for testing the ClientCache. I think it's a hard way to go about testing the transaction. Can you mock ClientCache instead?

File: client/python/tests/test_cache_map.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
+from cache_map import *
+
+import random
+
+TEST_CNT = 10000
```
Comment (Topher-the-Geek) [inline client/python/tests/test_cache_map.py]: This abbreviation is likely to be pronounced cunt. Let's spell out count. Thanks.

File: client/python/tests/test_client_cache.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+        cache = ClientCache(server, 
+            max_age=max_age)
+        result = cache.pool.request('GET', '/')
+        assert(result != None)
+
+        print "PASSED!"
+
+    # Test that calling read generates the expected request with headers
+    def test_client_cache_read(self):
```
Comment (Topher-the-Geek) [inline client/python/tests/test_client_cache.py]: I'm a bit confused about what the different test..read functions are each trying to achieve. I'd recommend that test names describe the input and expected behavior. Something like `Class_Inputs_Behavior`.

http://osherove.com/blog/2005/4/3/naming-standards-for-unit-tests.html


http://osherove.com/blog/2005/4/3/naming-standards-for-unit-tests.html

File: client/python/cache_map.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+            if (cache_key in cache_map):
+                cache_values = cache_map[cache_key]
+                cache_values.remove(cache_result)
+            # Reset the lru_list
+            lru_list.remove(node)
+        self.lru_list = lru_list
+
+    def _lru_remove(self, entries): 
+        lru_list = self.lru_list
+        for node in lru_list:
```
Comment (Topher-the-Geek) [inline client/python/cache_map.py]: This looks like it's O(N). Removing from a doubly-linked list should be O(1). LRU eviction should be O(1).

File: client/python/tx_clock.py (hunk 2 above)
Comment (jmfrye) [inline client/python/tx_clock.py]: Python long values don't have a maximum; I think I need to assign a max value for TxClock as a result:

http://stackoverflow.com/questions/9860588/maximum-value-for-long-integer

**Pull request closed without merging**